horizomer/utils/tree.py

```diff
@@ -13,6 +13,119 @@
 #
 
 from skbio import TreeNode
+from math import isclose
+from skbio.tree import MissingNodeError
+
+
+def _compare_length(node1, node2):
+    """Private function for compare_branch_lengths. Determines if lengths of the
+    two nodes are same.
+
+    Parameters
+    ----------
+    node1: skbio.TreeNode
+        first node to compare
+    node2: skbio.TreeNode
+        second node to compare
+
+    Returns
+    -------
+    bool
+        `True` if lengths of the two input nodes are None, same, or close
+        `False` otherwise
+
+    See Also
+    --------
+    compare_branch_length
+
+    Examples
+    --------
+    >>> from skbio import TreeNode
+    >>> tree = TreeNode.read(["(a:1,b:1)c;"])
+    >>> print(_compare_length(tree.find('a'), tree.find('b')))
+    True
+    >>> print(_compare_length(tree.find('a'), tree.find('c')))
+    False
+
+    """
+    if node1.length is None and node2.length is None:
+        return True
+    elif (node1.length is None) ^ (node2.length is None):
+        return False
+    elif isclose(node1.length, node2.length) is False:
+        return False
+    return True
+
+
+def compare_branch_lengths(tree1, tree2):
+    """Returns `True` if each corresponding node in 2 trees has same length.
+
+    Parameters
+    ----------
+    tree1: skbio.TreeNode
+        first tree to compare
+    tree2: skbio.TreeNode
+        second tree to compare
+
+    Returns
+    -------
+    bool
+        `True` if two input trees have same topologies and branch lengths
+        `False` otherwise
+
+    See Also
+    --------
+    compare_topology
+
+    Examples
+    --------
+    >>> from skbio import TreeNode
+    >>> tree1 = TreeNode.read(['((a:1, c:1):2, b:1);'])
+    >>> tree2 = TreeNode.read(['((a:1, c:1):2, b:1);'])
+    >>> print(compare_branch_lengths(tree1, tree2))
+    True
+    >>> tree3 = TreeNode.read(['((a:1, c:1):2, b);'])
+    >>> print(compare_branch_lengths(tree3, tree1))
+    False
+    """
+    tcopy1 = tree1.copy()
+    tcopy2 = tree2.copy()
+    stack = []  # stack to store nodes in tree2
+    count = 0
+
+    for node in tcopy1.postorder(include_self=False):
+        if node.is_tip():
+            try:
+                cur = tcopy2.find(node.name)
```

Comment:
why are you searching in the whole tree2, instead of only the corresponding sub-tree of tree2, which you obtain by simultaneously traverse tree1 and tree2, or better their corresponding siblings?

Reply:
In case that tree1 and tree2 have different node orderings but same topology, simultaneous traversal of tree1 and tree2 can be hardly achieved. For example:
tree1 = TreeNode.read(['((a:1,(b:1,c:1)d:1)e:1,f:1)g:1;'])
tree2 = TreeNode.read(['(f:1,((b:1,c:1)d:1,a:1)e:1)g:1;'])

While traversing tree1, the location of the corresponding node in tree2 is unknown, so find() is used here. find() creates a tip cache dictionary for the first time of method call, so it is not an expensive operation to search in whole tree 2.

Reply:
I agree that simultaneous traversal is not trivial. In fact, you are entering the realm of pairwise tree alignment if you try to solve this problem, which is at least O(n*m) for ordered trees with fixed maximum width and O(n^4) in more relaxed instances.
I wonder if you can rely on node names as anchor points to short-cut the tree alignment problem. If that is the case, you definitively need to elaborate on this major assumption in the docstring!